plugins/engines/ospray/ispc/render/SimulationRenderer.ispc

```diff
@@ -106,308 +106,253 @@ inline void composite(const varying vec4f& src, varying vec4f& dst,
-            //                }
-            //                const float giAttenuation =
-            //                    1.f - shadowIntensity * self->shadows;
-            //                voxelColor.x = voxelColor.x * giAttenuation;
-            //                voxelColor.y = voxelColor.y * giAttenuation;
-            //                voxelColor.z = voxelColor.z * giAttenuation;
-            //            }
-
-            // Compose final voxel color
-            vec4f voxelColor =
-                clamp(sampleOpacity / volume->samplingRate) *
-                make_vec4f(sampleColor.x, sampleColor.y, sampleColor.z, 1.0f);
-            composite(voxelColor, pathColor, (float)self->volumeSamplesPerRay);
+            volumeSampleColor = volume->transferFunction->getColorForValue(
+                volume->transferFunction, volumeSample);
+
+            // Voxel shading
+            bool shadingEnabled = volume->gradientShadingEnabled;
+            if (volume->singleShade)
+                shadingEnabled &= (samplingCounter == 0);
+            if (shadingEnabled)
+            {
+                vec3f gradient =
+                    safe_normalize(volume->computeGradient(volume, point));
+                // Setup differential geometry for the volume sample point.
+                DifferentialGeometry dg;
+                dg.P = point;
+                dg.epsilon = epsilon;
+                dg.Ns = gradient;
+
+                gradient =
+                    safe_normalize(volume->computeGradient(volume, point));
+
+                // Shading according to computed normal
+                const vec2f s = make_vec2f(0.5f);
+                const Light_SampleRes light =
+                    self->abstract.lights[0]->sample(self->abstract.lights[0],
+                                                     dg, s);
+                const float cosNL =
+                    (gradient.x == 0.f && gradient.y == 0.f &&
+                     gradient.z == 0.f)
+                        ? 1.f
+                        : abs(dot(safe_normalize(light.dir), gradient));
+
+                volumeSampleColor = volumeSampleColor * cosNL * light.weight;
+
+                // Specular
+                const vec3f reflectedNormal = normalize(
+                    ray.dir - 2.f * dot(ray.dir, gradient) * gradient);
+                const float angle =
+                    powf(max(0.f, dot(light.dir, reflectedNormal)), 20.f);
+                volumeSampleColor =
+                    volumeSampleColor + volume->specular * angle;
+            }
+
+            if (samplingCounter == 0 && self->shadows > 0.f)
+                // Compute shadow contributions
+                shadowIntensity =
+                    getVolumeShadowContributions(volume, self, ray, sample,
+                                                 point, epsilon);
         }
+        composite(make_vec4f(volumeSampleColor, sampleOpacity), pathColor,
+                  2.0f);
```

Comment:
Maybe this magic number should also be a constant

Reply:
DEFAULT_COMPOSITE_ALPHA_CORRECTION